### PR TITLE
feat: build EVM messages to allow sign Ethereum transactions

### PR DESCRIPTION
**Pull request by mmsqe**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -188,7 +199,13 @@ require (
...
 	gopkg.in/ini.v1 v1.67.0 // indirect
+	gopkg.in/natefinch/npipe.v2 v2.0.0-20160621034901-c1b8fa8bdcce // indirect
 	nhooyr.io/websocket v1.8.6 // indirect
 	pgregory.net/rapid v0.5.5 // indirect
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
+
+replace (
+	github.com/ChainSafe/go-schnorrkel => github.com/ChainSafe/go-schnorrkel v0.0.0-20200405005733-88cbf1b4c40d
+	github.com/evmos/ethermint => github.com/crypto-org-chain/ethermint v0.0.0-20230801014829-2df053357655
```
Comment (yihuang) [inline go.mod:210]: does `crypto-org-chain/ethermint` contains some necessary changes for this?

Comment (mmsqe) [reply go.mod:210]: need depends on sdk47

Comment (yihuang) [reply go.mod:210]: I see

File: relayer/chains/cosmos/tx.go [hunk 2]
```diff
@@ -649,22 +669,260 @@ func (cc *CosmosProvider) buildMessages(
...
+	ctx context.Context,
+	txf *tx.Factory,
+	txb client.TxBuilder,
+	txSignerKey string,
+) ([]byte, uint64, sdk.Coins, error) {
+	if err := tx.Sign(*txf, txSignerKey, txb, false); err != nil {
+		return nil, 0, sdk.Coins{}, err
+	}
+	tx := txb.GetTx()
+	txBytes, err := cc.signTxBytes(ctx, tx)
```
Comment (yihuang) [inline relayer/chains/cosmos/tx.go]: it seems signed twice here?

EDIT: I think it's a method name confusion.

File: relayer/chains/cosmos/tx.go [hunk 3]
```diff
@@ -649,22 +669,260 @@ func (cc *CosmosProvider) buildMessages(
...
+}
 
+func (cc *CosmosProvider) signTxBytes(
+	ctx context.Context,
+	tx authsigning.Tx,
+) ([]byte, error) {
+	var txBytes []byte
 	// Generate the transaction bytes
-	txBytes, err = cc.Cdc.TxConfig.TxEncoder()(tx)
+	if err := retry.Do(func() error {
```
Comment (yihuang) [inline relayer/chains/cosmos/tx.go]: why do we need retry, I think encoding the tx is a pure function.

File: relayer/chains/cosmos/tx.go [hunk 4]
```diff
@@ -649,22 +669,260 @@ func (cc *CosmosProvider) buildMessages(
...
-	fees = tx.GetFee()
+func convertCoins(prices string) (sdk.Coins, error) {
+	coins, err := sdk.ParseCoinsNormalized(prices)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "failed to parse coins")
+	}
+	return coins, nil
+}
 
+func (cc *CosmosProvider) signTxBytes(
```
Comment (yihuang) [inline relayer/chains/cosmos/tx.go]: this method encode the signed tx, not signing the tx?

Comment (mmsqe) [reply relayer/chains/cosmos/tx.go]: yes

File: relayer/chains/cosmos/tx.go (hunk 2 above)
Comment (mmsqe) [reply relayer/chains/cosmos/tx.go]: yes, only call once for buildMessages or buildEvmMessages

File: relayer/chains/cosmos/tx.go [hunk 5]
```diff
@@ -649,22 +669,248 @@ func (cc *CosmosProvider) buildMessages(
...
+
+func convertCoins(prices string) (sdk.Coins, error) {
+	coins, err := sdk.ParseCoinsNormalized(prices)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "failed to parse coins")
+	}
+	return coins, nil
+}
 
+func (cc *CosmosProvider) encodeSignedTx(
```
Comment (yihuang) [inline relayer/chains/cosmos/tx.go]: do we need this method at all, it's an one-liner.

**Review by yihuang — APPROVED** [review]

File: relayer/chains/cosmos/tx.go (hunk 5 above)
Comment (mmsqe) [reply relayer/chains/cosmos/tx.go]: removed

**Pull request merged**